src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -51,6 +58,41 @@ internal partial class DiagnosticIncrementalAnalyzer
             {
                 throw ExceptionUtilities.Unreachable;
             }
+
+            static bool IsAnalyzerSuppressedForDocument(
+                DiagnosticAnalyzer analyzer,
+                BackgroundAnalysisScope analysisScope,
+                bool isActiveDocument,
+                bool isOpenDocument,
+                bool isGeneratedRazorDocument)
+            {
+                if (isGeneratedRazorDocument)
+                {
+                    // This is a generated Razor document, and they always want all analyzer diagnostics.
+                    return false;
+                }
+
+                var analyzerEnabled = analysisScope switch
+                {
+                    // Analyzers are only enabled for active document.
+                    // Compiler analyzer is special and enabled for all open documents.
+                    BackgroundAnalysisScope.ActiveFile => analyzer.IsCompilerAnalyzer() ? isOpenDocument : isActiveDocument,
+
+                    // Analyzers are disabled for all documents.
+                    // Compiler analyzer is special and enabled for all open documents.
+                    BackgroundAnalysisScope.None => isOpenDocument && analyzer.IsCompilerAnalyzer(),
+
+                    // All analyzers, including compiler analyzer, are enabled for all open documents.
+                    BackgroundAnalysisScope.OpenFiles => isOpenDocument,
+
+                    // All analyzers, including compiler analyzer, are enabled for all documents.
+                    BackgroundAnalysisScope.FullSolution => true,
```

Comment:
note: first: this entire switch exemplifies why i think this UI feature is just confusing.  

Second: i think a much cleaner way to support this *internally* is to map the user UI feature to a set of two *internal* flags.  one controlling the compiler analyzer, and one controlling any other analyzer.  then, all these checks get so much simpler.  We just check if tis' the compilr analyzer or not, and then we we do the appropropriate check of each analysis scope option.

Reply:
Simplified with https://github.com/dotnet/roslyn/pull/57172/commits/1c6887ccbd5cc63be4cb66443c1d5dabfa4fcc2c